### PR TITLE
Enhanced find method

### PR DESCRIPTION
**Pull request by PROGRAMMERHAO**
1. Add support for finding events by either from or to date, as well as finding deadlines.
2. Add checks so that user cannot enter prefixes with empty values when using find command.
3. Allow user to specify whether the tags they are looking for must all be present in a task or not.
4. Allow user to search for multiple segments of one task's description or name at the same time using "all/".

### CODE REVIEW

File: src/test/java/seedu/task/logic/parser/FindCommandParserTest.java [hunk 1]
```diff
@@ -27,16 +34,29 @@ public void parse_validArgs_returnsFindCommand() {
...
         assertParseSuccess(parser, " n/Alice Bob", expectedFindCommand);
+        List<String> names = new ArrayList<>();
+        names.add("Bob");
+        names.add("Alice");
         FindCommand expectedFindCommand2 =
-                new FindCommand(new NameContainsKeywordsPredicate("Alice \n \t Bob"));
-        // multiple whitespaces between keywords
-        assertParseSuccess(parser, " \n n/Alice \n \t Bob  \t", expectedFindCommand2);
+            new FindCommand(new NameContainsAllKeywordsPredicate(names));
+        assertParseSuccess(parser, " all/hi n/Alice n/Bob", expectedFindCommand2);
```
Comment (Huggenguggen) [inline src/test/java/seedu/task/logic/parser/FindCommandParserTest.java:42]: Something for consideration, all/ should not allow words after it as it should be just used for flagging.

File: src/main/java/seedu/task/logic/parser/FindCommandParser.java [hunk 2]
```diff
@@ -32,22 +45,48 @@ public class FindCommandParser implements Parser<FindCommand> {
...
+            return new FindCommand(new DescContainsKeywordsPredicate(description.toString()));
         }
         if (parseTagsForFind(argMultimap.getAllValues(PREFIX_TAG)).isPresent()) {
-            return new FindCommand(new TagsContainsKeywordsPredicate(argMultimap.getAllValues(PREFIX_TAG)));
+            Set<Tag> tags = ParserUtil.parseTags(argMultimap.getAllValues(PREFIX_TAG));
+            if (argMultimap.getValue(PREFIX_ALLMATCH).isPresent()) {
+                return new FindCommand(new TagsContainsAllKeywordsPredicate(ParserUtil.parseTagsToList(tags)));
+            }
+            return new FindCommand(new TagsContainsKeywordsPredicate(ParserUtil.parseTagsToList(tags)));
+        }
```
Comment (Huggenguggen) [inline src/main/java/seedu/task/logic/parser/FindCommandParser.java:77]: There is quite a lot of code duplication here, possible chance to abstract out to another method? Perhaps it takes in a prefix and returns a find command.

File: src/main/java/seedu/task/logic/parser/FindCommandParser.java [hunk 3]
```diff
@@ -32,22 +45,48 @@ public class FindCommandParser implements Parser<FindCommand> {
...
+            String date = Date.parseFindDate(argMultimap.getValue(PREFIX_DEADLINE).get());
+            return new FindCommand(new DeadlineDateContainsKeywordsPredicate(date));
+        }
+        if (argMultimap.getValue(PREFIX_FROM).isPresent()) {
+            String date = Date.parseFindDate(argMultimap.getValue(PREFIX_FROM).get());
+            return new FindCommand(new EventFromContainsKeywordsPredicate(date));
+        }
+        if (argMultimap.getValue(PREFIX_TO).isPresent()) {
+            String date = Date.parseFindDate(argMultimap.getValue(PREFIX_TO).get());
+            return new FindCommand(new EventToContainsKeywordsPredicate(date));
```
Comment (Huggenguggen) [inline src/main/java/seedu/task/logic/parser/FindCommandParser.java:88]: Similar to above, a possible candidate for abstraction/refactoring

File: src/test/java/seedu/task/logic/parser/FindCommandParserTest.java (hunk 1 above)
Comment (PROGRAMMERHAO) [reply src/test/java/seedu/task/logic/parser/FindCommandParserTest.java]: for now this is the behaviour but we will change in the future

File: src/main/java/seedu/task/logic/parser/FindCommandParser.java (hunk 3 above)
Comment (PROGRAMMERHAO) [reply src/main/java/seedu/task/logic/parser/FindCommandParser.java]: ok done!

File: src/main/java/seedu/task/logic/parser/FindCommandParser.java (hunk 2 above)
Comment (PROGRAMMERHAO) [reply src/main/java/seedu/task/logic/parser/FindCommandParser.java]: created a new parseForFindCommand method to handle this

**Pull request merged**